### ISSUE TITLE
False positive for RedundantCondition ?

### ISSUE THREAD

**Issue by chapa** [vimeo/psalm#3378]
Hi,

I think there may be a false positive with RedundantCondition in `do ... while` conditions : https://psalm.dev/r/573f17e1e9

If I take out the variable `$this->continuePolling` from the `while` condition ...
```php
do {
    $event = $this->fetchEvent();

    if ($this->shouldReturnEvent($event)) {
        return $event;
    }
} while ($this->continuePolling && null !== $event);
```
... and put it in a if just before ...
```php
do {
    $event = $this->innerStream->nextEvent();

    if ($this->shouldReturnEvent($event)) {
        return $event;
    }

    if (!$this->continuePolling) {
        break;
    }
} while (null !== $event);
```
... there is no more error : https://psalm.dev/r/9d00aadfb1

**Comment by muglug** [comment]
Simplified to

```php
class Event {}

function fetchEvent(): ?Event {
    return rand(0, 1) ? new Event() : null;
}

function nextEvent(bool $c): void {
    do {
        $e = fetchEvent();
    } while ($c && $e);
}
```

**Comment by chapa** [comment]
Thank you ! :+1:
